pkg/cmd/pr/cleanup/cleanup.go

```diff
@@ -0,0 +1,393 @@
+	cmd.Flags().BoolVarP(&opts.Strict, "strict", "", false, "Both --exclude-closed and --exclude-behind")
+	cmd.Flags().BoolVarP(&opts.MergedOnly, "exclude-closed", "", false, "Exclude branches of closed pull requests")
+	cmd.Flags().BoolVarP(&opts.UpToDateOnly, "exclude-behind", "", false, "Exclude branches that are behind their remote")
+	cmd.Flags().BoolVarP(&opts.Yes, "yes", "", false, "Skip deletion confirmation")
+
+	return cmd
+}
+
+func cleanupRun(opts *CleanupOptions) error {
+	// Validate input arguments: --all and PR selector are mutually exclusive, but
+	// at least one must be set.
+	if opts.All && opts.SelectorArg != "" {
+		return errors.New("Invalid arguments: cannot set both PR and --all")
+	} else if opts.SelectorArg == "" && !opts.All {
+		return errors.New("Invalid arguments: must set either PR or --all")
+	}
+
+	// Set flags.
+	if opts.Strict {
+		opts.MergedOnly = true
+		opts.UpToDateOnly = true
+	}
+
+	if opts.All {
+		// Get all local branches and their upstreams.
+		ctx := context.Background()
+		localBranches := opts.GitClient.LocalBranches(ctx)
+		var branchesWithUpstream []git.Branch
+		for _, localBranch := range localBranches {
+			if localBranch.Upstream.RemoteName == "" {
+				continue
+			}
+			branchesWithUpstream = append(branchesWithUpstream, localBranch)
+		}
+
+		// Get PRs associated with upstream branches.
+		opts.IO.StartProgressIndicatorWithLabel(
+			fmt.Sprintf(
+				"Loading PRs for %d local branches with upstreams. This can take a minute if you have many branches...\n",
+				len(branchesWithUpstream),
+			),
+		)
+
+		var prs []*api.PullRequest
+		prStates := []string{"MERGED", "CLOSED"}
+		if opts.MergedOnly {
+			prStates = []string{"MERGED"}
+		}
+		for _, branch := range branchesWithUpstream {
+			// TODO: Can these be loaded in parallel?
+			// TODO: This causes the progress indicator to "reset" very frequently.
+			// Should the Finder itself have a progress indicator? Perhaps we should
+			// invert that so consumers have control of the indicator instead.
+			pr, _, err := opts.Finder.Find(shared.FindOptions{
+				Selector: branch.Upstream.BranchName,
+				Fields:   []string{"headRefOid", "headRefName", "title", "state"},
+				States:   prStates,
+			})
+			if _, ok := err.(*shared.NotFoundError); ok {
+				continue
+			}
+			if err != nil {
+				return err
+			} else {
+				prs = append(prs, pr)
+			}
+		}
+		opts.IO.StopProgressIndicator()
+
+		// Reorganize branches for fast lookup.
+		branchesByRemoteBranchName := make(map[string]git.Branch)
+		for _, branch := range branchesWithUpstream {
+			branchesByRemoteBranchName[branch.Upstream.BranchName] = branch
+		}
+
+		// Get the list of candidate branch deletions.
+		//
+		// Any local branch whose HEAD is a commit of a merged or closed PR is a
+		// candidate for deletion, because the local branch's history is a prefix of
+		// the remote branch's history (i.e. there are no local commits that the
+		// upstream does not have).
```

Comment:
:thinking: So for a squash- or rebase-merged PR, this logic verifies that the user’s local commits made it into the PR before it was merged, and we trust that the resulting commit to the base branch actually reflected those changes (perhaps with some amendments from the maintainer). Seems close enough. Compare discussion of `git branch -d` behavior in #380.